server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -144,111 +145,104 @@ public RepositoriesService(
-                    } else {
-                        logger.info("put repository [{}]", request.name());
+                        @Override
+                        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+                            if (changed) {
+                                if (found) {
+                                    logger.info("updated repository [{}]", request.name());
+                                } else {
+                                    logger.info("put repository [{}]", request.name());
+                                }
+                            }
+                            publicationStep.onResponse(oldState != newState);
+                        }
                     }
+                );
+                publicationStep.addListener(clusterUpdateStep.delegateFailureAndWrap((ignored1, changed) -> {
+                    acknowledgementStep.addListener(clusterUpdateStep.delegateFailureAndWrap((ignored2, ack) -> {
+                        clusterUpdateStep.onResponse(new RegisterRepositoryTaskResult(ack, changed));
+                    }));
+                }));
+            })
+            .<AcknowledgedResponse>andThen((verificationStep, taskResult) -> {
+                if (request.verify() == false) {
+                    verificationStep.onResponse(taskResult.ackResponse);
+                } else {
+                    SubscribableListener
+
+                        .<List<DiscoveryNode>>newForked(verifyRepositoryStep -> {
+                            if (taskResult.ackResponse.isAcknowledged() && taskResult.changed) {
+                                verifyRepository(request.name(), verifyRepositoryStep);
+                            } else {
+                                verifyRepositoryStep.onResponse(null);
+                            }
+                        })
+                        // When verification has completed, get the repository data for the first time
+                        .<RepositoryData>andThen((getRepositoryDataStep, ignored) -> {
+                            threadPool.generic()
+                                .execute(
+                                    ActionRunnable.wrap(
+                                        getRepositoryDataStep,
+                                        ll -> repository(request.name()).getRepositoryData(
+                                            // TODO contemplate threading, do we need to fork, see #101445?
+                                            EsExecutors.DIRECT_EXECUTOR_SERVICE,
+                                            ll
+                                        )
+                                    )
+                                );
+                        })
+                        // When the repository metadata is ready, update the repository UUID stored in the cluster state, if available
+                        .<Void>andThen((updateRepoUuidStep, repositoryData) -> {
```

Comment:
Nit: unnecessary `{...}` block, this can just be an expression lambda.

```suggestion
                        .<Void>andThen((updateRepoUuidStep, repositoryData) -> 
```